hypothesis-python/src/hypothesis/internal/scrutineer.py

```diff
@@ -38,11 +39,45 @@ def trace(self, frame, event, arg):
         if event == "call":
             return self.trace
         elif event == "line":
-            fname = frame.f_code.co_filename
-            if should_trace_file(fname):
-                current_location = (fname, frame.f_lineno)
-                self.branches.add((self._previous_location, current_location))
-                self._previous_location = current_location
+            self.trace_line(frame.f_code, frame.f_lineno)
+
+    def trace_line(self, code: types.CodeType, line_number: int):
+        fname = code.co_filename
+        if should_trace_file(fname):
+            current_location = (fname, line_number)
+            self.branches.add((self._previous_location, current_location))
+            self._previous_location = current_location
+
+
+# where possible, we'll use 3.12's new sys.monitoring module for low-overhead
+# coverage instrumentation. Otherwise, we'll default to sys.settrace.
+# This can be simplified once we drop 3.11.
+# tool_id = 2 is designated for coverage.
+MONITORING_TOOL_ID = 2
+if sys.version_info[:2] >= (3, 12):
+    MONITORING_EVENTS = {sys.monitoring.events.LINE: "trace_line"}
+
+
+def settracer(tracer: Tracer):
+    if sys.version_info[:2] < (3, 12):
+        sys.settrace(tracer.trace)
+        return
+
+    sys.monitoring.use_tool_id(MONITORING_TOOL_ID, "scrutineer")
+    for event, callback_name in MONITORING_EVENTS.items():
+        sys.monitoring.set_events(MONITORING_TOOL_ID, event)
+        callback = getattr(tracer, callback_name)
+        sys.monitoring.register_callback(MONITORING_TOOL_ID, event, callback)
+
+
+def removetracer():
+    if sys.version_info[:2] < (3, 12):
+        sys.settrace(None)
+        return
+
+    sys.monitoring.free_tool_id(MONITORING_TOOL_ID)
```

Comment:
PEP 669 claims that `use_tool_id`, `free_tool_id`, and `set_events` "should be regarded as slow". This PR currently calls all of these once per test execution. I haven't profiled these, but if this has an impact, we could consider registering once at the global level instead, and enable/disable the callbacks per test execution instead.

Reply:
Hmm.  I think we should just ignore this for now and do the simple thing, and we can complicate things iff profiling shows we need to later.